File: BEmu/MarketDataRequest/ElementMarketString.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="MarketDataRequest/ElementMarketString.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementMarketString : Element
    {
        private readonly string _name, _value;

        internal ElementMarketString(string name, string value)
        {
            this._name = name;
            this._value = value;
        }

        public override Name Name { get { return new Name(this._name); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return 0; } }
        public override bool IsNull { get { return false; } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return false; } }

        public override object this[int index]
        {
            get
            {
                if (index == 0)
                    return this._value;
                else
                    throw new IndexOutOfRangeException();
            }
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            return base.PrettyPrintHelper(tabIndent, this._value);
        }

        public override object GetValue()
        {
            return this._value;
        }

        public override string GetValueAsString()
        {
            return this._value;
        }
    }
}

File: BEmu_cpp/src/MarketDataRequest/ElementMarketString.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/MarketDataRequest/ElementMarketString.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "MarketDataRequest/ElementMarketString.h"

namespace BEmu
{
	namespace MarketDataRequest
	{
		ElementMarketString::ElementMarketString(std::string name, std::string value)
		{
			this->_name = name;
			this->_value = value;
		}

		Name ElementMarketString::name() const
		{
			return Name(this->_name.c_str());
		}

		size_t ElementMarketString::numValues() const
		{
			return 1;
		}

		size_t ElementMarketString::numElements() const
		{
			return 0;
		}

		bool ElementMarketString::isNull() const
		{
			return false;
		}

		bool ElementMarketString::isArray() const
		{
			return false;
		}

		bool ElementMarketString::isComplexType() const
		{
			return false;
		}

		const char * ElementMarketString::getValueAsString(int index) const
		{
			if(index == 0)
				return this->toCharPointer(this->_value);
			else
				throw elementPtrEx;
		}

		std::ostream& ElementMarketString::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			this->prettyPrintHelper(stream, level, spacesPerLevel, this->_value);
			return stream;
		}

	}
}